### PR TITLE
inject: Set LINKERD2_PROXY_CORES from the cpu limit

### PR DESCRIPTION
**Pull request by olix0r**
Per #5165, Kubernetes does not necessarily limit the proxy's access to
cores via `cgroups` when a CPU limit is set. As of #5168, the proxy now
supports a `LINKERD2_PROXY_CORES` environment configuration.

This change modifies the proxy injector to ensure that this environment
is configured from `Values.proxy.cores` Helm value and the
`config.linkerd.io/proxy-cpu-limit` annotation.

### CODE REVIEW

**Comment by Pothulapati** [comment]
Hmmm, Why not use `values.Global.Proxy.Resources.CPU.Limit` for the same? should make this easy as if we use the same field to set the new env in the template, all the injection (from annotation) and install flow (i.e with `--proxy-cpu-limit)` would work right away. WDYT?

 With this route all of them have to be updated to take into account this new field.

**Comment by alpeb** [comment]
@Pothulapati I guess that's because you wouldn't be able to implement directly in the template what `toWholeCPUCores` does...

**Comment by Pothulapati** [comment]
@alpeb Ahh, That makes sense then! I'm also guessing it would be really hard to convert k8s.quantity into a whole number in Helm.

With this new variable, we should be overwriting this variable wherever `.values.Global.Proxy.Resources.CPU.Limit` is being to make inject and install overriding work.

**Comment by alpeb** [comment]
> With this new variable, we should be overwriting this variable wherever `.values.Global.Proxy.Resources.CPU.Limit` is being to make inject and install overriding work.

Right, for instance `linkerd install --proxy-cpu-limit` is not affecting this new env var, not for control-plane pods nor for pods injected by the injector.

This also made me realize that for the helm install workflows, control-plane pods won't get this env var even if `global.Proxy.Resources.CPU.Limit` is explicitly set. So we end up with the same problem for CP proxies when the node has more cores than that Limit...

One not so straightforward way around that would be to expose in `values.yaml` separate `.proxy.core` entries for each CP component with anchors pointing to `global.Proxy.Resources.CPU.Limit` but then the proxy would have to deal with non-integer values. Actually if the proxy could deal with k8's Quantities directly, we could feed the env var the Limit directly like @Pothulapati suggested and all these issues would go away without additional `values.yaml` entries...

**Comment by olix0r** [comment]
> With this new variable, we should be overwriting this variable wherever .values.Global.Proxy.Resources.CPU.Limit is being to make inject and install overriding work.

It's my preference that we not delay the edge release on this (and that we not make further changes between the edge and stable release, unless critical). I think it's sufficient to document this behavior, as it's only necessary when configuring via Helm and kubelet is not enforcing QoS.

The one thing we should confirm, though, is that the correct values are set when the `--proxy-cpu-limit` flag is set at install/upgrade-time.

File: cli/cmd/options.go [hunk 1]
```diff
@@ -66,10 +66,11 @@ func makeInstallUpgradeFlags(defaults *l5dcharts.Values) ([]flag.Flag, *pflag.Fl
 					// individual flags can override the HA defaults.  This means that the HA flag must appear
 					// before the individual flags in the slice passed to flags.ApplyIfSet.
 					values.ControllerReplicas = haValues.ControllerReplicas
-					values.Global.Proxy.Cores = haValues.Global.Proxy.Cores
 					values.Global.Proxy.Resources.CPU.Request = haValues.Global.Proxy.Resources.CPU.Request
 					values.Global.Proxy.Resources.Memory.Request = haValues.Global.Proxy.Resources.Memory.Request
-					values.Global.Proxy.Resources.CPU.Limit = haValues.Global.Proxy.Resources.CPU.Limit
+					// NOTE: CPU Limits are not currently set by default HA charts.
+					//values.Global.Proxy.Cores = haValues.Global.Proxy.Cores
+					//values.Global.Proxy.Resources.CPU.Limit = haValues.Global.Proxy.Resources.CPU.Limit
```
Comment (Pothulapati) [inline cli/cmd/options.go:73]: wondering what difference this makes as we are storing empty values of these variables even without this change? :thinking:

**Review by Pothulapati — APPROVED** [review]
Install and Inject work as expected i.e new env is also set along with the resources.

File: cli/cmd/options.go (hunk 1 above)
Comment (olix0r) [reply cli/cmd/options.go:73]: AIUI, this prevents us from overwriting values set by `--proxy-cpu-limit` when it is set before `--ha`

File: pkg/inject/inject_test.go [hunk 2]
```diff
@@ -288,3 +291,27 @@ func TestGetOverriddenValues(t *testing.T) {
...
+
+func TestWholeCPUCores(t *testing.T) {
+	for _, c := range []struct {
+		v string
+		n int
+	}{
+		{v: "1", n: 1},
+		{v: "1m", n: 1},
+		{v: "1000m", n: 1},
+		{v: "1001m", n: 2},
```
Comment (Pothulapati) [inline pkg/inject/inject_test.go:303]: nit: why not round instead? so that this is `1` and not `2` so that we are not going out of the boundary :thinking:

Comment (olix0r) [reply pkg/inject/inject_test.go:303]: (1) quantity exposes a `RoundUp` and not `Round`
(2) i think if you say the proxy can use more than a core and we only allocate a single core, we'll never use the full limit, so rounding up better-reflects the user's intent, I think

**Review by alpeb — APPROVED** [review]
:shipit:

**Pull request merged**